src/main/kotlin/com/team4099/robot2022/subsystems/Drivetrain.kt

```diff
@@ -337,12 +335,9 @@ object Drivetrain : SubsystemBase() {
     zeroDrive()
   }
 
-  fun zeroGyro() {
-    gyro.angleAdjustment = gyroChange.inDegrees
-  }
-  fun zeroGyro(offset: Angle) {
-    zeroGyro()
-    gyroOffset = offset
+  fun zeroGyro(toAngle: Angle = 0.degrees) {
+    gyro.angleAdjustment = -(gyroAngle).inDegrees
+    gyroOffset = toAngle
```

Comment:
what happens here if you zero twice with different offsets? this is actually likely to happen in a match if you start sideways and then gyro drift causes the driver to manually rezero.

to me the easiest solution is to just forget about the navx built in angle adjustment, and put everything into `gyroOffset`. basically what you want is to set `gyroOffset` to an angle that, when added to `gyro.angle`, gives you `toAngle`

Reply:
wait also hold on why don't we use this for pose? if you set the drivetrain pose you should automatically zero the gyro to the angle from that pose so that if you start sideways in auto it auto zeros correctly (simple custom setter for pose i think)

Reply:
if we do that for pose then we need to do a similar toAngle thing for both zeroSteering and zeroDrive. Also SwerveDriveOdometry states that it automatically updates the gyro but idk if that's truely the case.

Reply:
so idk why we would need that for zeroSteering, which just grabs the absolute encoder position and uses it to reset the steering motor's internal encoder (what would an offset mean for this?).

for zeroDrive, we pretty much never use the position reading from the drive encoders so this is actually pretty useless. the SwerveDriveOdometry class uses the angle of the module + the velocity its going at to compute a velocity for the robot, and it combines that with the time since last update to estimate a position delta. so it's critical that the initial position be correct -- which we do ensure by resetting the pose at the start of auto.

what I was saying with the gyro is that currently, the only way we store the info on the starting pose of the robot (position and angle both) is in the starting pose of the first trajectory we run in auto. so while its useful to have this zeroGyro method we don't actually call it in any useful place rn. so instead, when we manually set the pose (at the beginning of the auto code), we take that opportunity to also configure the gyro offset correctly, it seems to make sense to do it at that time

Reply:
or in other words: when we do `swerveDriveOdometry.resetPosition(value.pose2d, gyroAngle.inRotation2ds)` in the setter, we are basically saying we're taking the Pose2d from the pose we're fed, but we're not resetting the angle. there doesn't seem to be any reason we shouldn't just reset the angle here.

Reply:
SwerveDriveOdometry doesn't update the gyro afaik, it uses the gyro position to correctly calculate odometry. but the issue is if you give it a bad gyro angle (which is what happens if you don't reset it) then it can't correctly calculate the odometry